src/translator/tag_processor.h

```diff
@@ -0,0 +1,130 @@
+#ifndef BERGAMOT_TRANSLATOR_TAG_NESTING_H_
+#define BERGAMOT_TRANSLATOR_TAG_NESTING_H_
+
+#include <vector>
+#include <string>
+
+#include "data/alignment.h"
+#include "annotation.h"
+
+namespace marian {
+namespace bergamot {
+
+struct TagNode {
+  size_t parent;  // can be removed; used if backtrack
+  ByteRange bound;
+  std::string_view label;
+  std::vector<size_t> child;
+
+  TagNode(ByteRange bound,std::vector<size_t> child): bound(bound),child(std::move(child)){};
+};
+
+class TagProcessor {
+  std::vector<TagNode> tagTree_;
+
+  size_t srcLength_;
+  size_t tgtLength_;
+
+  std::vector<std::vector<std::vector<double>>> inside_;
+
+public:
+  std::vector<TagNode> tagTreeTarget; // store query results
+
+  TagProcessor(const marian::data::SoftAlignment &align,
+               std::vector<TagNode> tagTree,
+               size_t srcLength,
+               size_t tgtLength) : tagTree_(std::move(tagTree)), srcLength_(srcLength), tgtLength_(tgtLength) {
+    inside_.resize(srcLength, std::vector<std::vector<double>>(srcLength, std::vector<double>(tgtLength, 0)));
```

Comment:
Allocate `srcLength * srcLength * tgtLength` on a flat array and use [address-calculation](https://en.wikipedia.org/wiki/Row-_and_column-major_order#Address_calculation_in_general)?

Reply:
That optimisation is good to be done later. Let me first make the core algorithm correct.